### ISSUE TITLE
Problem deploying to Clojars - related to test artifacts?

### ISSUE THREAD

**Issue by mikera** [clojars/clojars-web#640]
I hit a strange error when deploying `core.matrix` to Clojars:

```
    [INFO] --- maven-install-plugin:2.4:install (default-install) @ core.matrix ---
    [INFO] Installing C:\Users\Mike\git\core.matrix\target\checkout\target\core.matrix-0.60.1.jar to K:\Prog\m2\repository\net\mikera\core.matrix\0.60.1\core.matrix-0.60.1.jar
    [INFO] Installing C:\Users\Mike\git\core.matrix\target\checkout\pom.xml to K:\Prog\m2\repository\net\mikera\core.matrix\0.60.1\core.matrix-0.60.1.pom
    [INFO] Installing C:\Users\Mike\git\core.matrix\target\checkout\target\core.matrix-0.60.1-sources.jar to K:\Prog\m2\repository\net\mikera\core.matrix\0.60.1\core.matrix-0.60.1-sources.jar
    [INFO] Installing C:\Users\Mike\git\core.matrix\target\checkout\target\core.matrix-0.60.1-javadoc.jar to K:\Prog\m2\repository\net\mikera\core.matrix\0.60.1\core.matrix-0.60.1-javadoc.jar
    [INFO] Installing C:\Users\Mike\git\core.matrix\target\checkout\target\core.matrix-0.60.1-tests.jar to K:\Prog\m2\repository\net\mikera\core.matrix\0.60.1\core.matrix-0.60.1-tests.jar
    [INFO] Installing C:\Users\Mike\git\core.matrix\target\checkout\target\core.matrix-0.60.1.jar.asc to K:\Prog\m2\repository\net\mikera\core.matrix\0.60.1\core.matrix-0.60.1.jar.asc
    [INFO] Installing C:\Users\Mike\git\core.matrix\target\checkout\target\core.matrix-0.60.1.pom.asc to K:\Prog\m2\repository\net\mikera\core.matrix\0.60.1\core.matrix-0.60.1.pom.asc
    [INFO] Installing C:\Users\Mike\git\core.matrix\target\checkout\target\core.matrix-0.60.1-sources.jar.asc to K:\Prog\m2\repository\net\mikera\core.matrix\0.60.1\core.matrix-0.60.1-sources.jar.asc
    [INFO] Installing C:\Users\Mike\git\core.matrix\target\checkout\target\core.matrix-0.60.1-javadoc.jar.asc to K:\Prog\m2\repository\net\mikera\core.matrix\0.60.1\core.matrix-0.60.1-javadoc.jar.asc
    [INFO] Installing C:\Users\Mike\git\core.matrix\target\checkout\target\core.matrix-0.60.1-tests.jar.asc to K:\Prog\m2\repository\net\mikera\core.matrix\0.60.1\core.matrix-0.60.1-tests.jar.asc
    [INFO]
    [INFO] --- maven-deploy-plugin:2.7:deploy (default-deploy) @ core.matrix ---
    Uploading: https://clojars.org/repo/net/mikera/core.matrix/0.60.1/core.matrix-0.60.1.jar
    Uploaded: https://clojars.org/repo/net/mikera/core.matrix/0.60.1/core.matrix-0.60.1.jar (130 kB at 24 kB/s)
    Uploading: https://clojars.org/repo/net/mikera/core.matrix/0.60.1/core.matrix-0.60.1.pom
    Uploaded: https://clojars.org/repo/net/mikera/core.matrix/0.60.1/core.matrix-0.60.1.pom (5.9 kB at 1.9 kB/s)
    Downloading: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml
    Downloaded: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml (2.9 kB at 5.5 kB/s)
    Downloading: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml
    Downloaded: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml (2.9 kB at 5.8 kB/s)
    Uploading: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml
    Uploaded: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml (2.9 kB at 937 B/s)
    Uploading: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml
    [INFO] ------------------------------------------------------------------------
    [INFO] BUILD FAILURE
    [INFO] ------------------------------------------------------------------------
    [INFO] Total time: 55.946 s
    [INFO] Finished at: 2017-05-16T09:43:10+08:00
    [INFO] Final Memory: 31M/366M
    [INFO] ------------------------------------------------------------------------
    [ERROR] Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:2.7:deploy (default-deploy) on project core.matrix: Failed to deploy metadata: Could not transfer metadata net.mikera:core.matrix/maven-metadata.xml from/to clojars.org (https://clojars.org/repo): Access denied to: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml, ReasonPhrase: Forbidden - redeploying non-snapshots is not allowed (see https://git.io/v1IAs). -> [Help 1]
    [ERROR]
    [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
    [ERROR] Re-run Maven using the -X switch to enable full debug logging.
    [ERROR]
    [ERROR] For more information about the errors and possible solutions, please read the following articles:
    [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 03:32 min
[INFO] Finished at: 2017-05-16T09:43:10+08:00
[INFO] Final Memory: 13M/232M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-release-plugin:2.5.2:perform (default-cli) on project core.matrix: Maven execution failed, exit code: '1' -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

I suspect it might be due to the presence of the test artifacts along with the main jar, since that might look like an attempt to redeploy the same artifact. I have no way to confirm this, but it does appear that the test jars are *not* making it into the release. I've tried with 0.60.0 and 0.60.1.

Any ideas? I've been deploying this setup successfully to Clojars in the past

**Comment by mikera** [comment]
Quick help appreciated with this BTW - it's breaking builds for downstream users who depend on the tests!

**Comment by tobias** [comment]
This may be a regression around how we handle artifacts uploaded after `maven-metadata.xml` (which is where maven uploads artifacts with classifiers, so tests, source, etc). I suspect this was caused by changes I made last week to handle multiple deployments sharing the same cookie store. I'll take a look now.

**Comment by tobias** [comment]
I can't recreate this locally, but I don't see the double upload of `maven-metadata.xml` locally either. What version of `mvn` are you using? I'm testing with 3.3.9.

**Comment by mikera** [comment]
I was using 3.5.0 recent install

**Comment by tobias** [comment]
Thanks. I can recreate it with 3.5.0. It looks like they've changed the deploy process a bit. Working on a fix now.

**Comment by tobias** [comment]
I've deployed a fix. Can you deploy `core.matrix` again and confirm it works for you? You'll have to bump the version - you won't be able to deploy over one of the failed versions.

**Comment by mikera** [comment]
Still failed when I attempted to deploy `0.60.2`

Relevant messages:
```
    [INFO] --- maven-deploy-plugin:2.7:deploy (default-deploy) @ core.matrix ---
    Uploading: https://clojars.org/repo/net/mikera/core.matrix/0.60.2/core.matrix-0.60.2.jar
    Uploaded: https://clojars.org/repo/net/mikera/core.matrix/0.60.2/core.matrix-0.60.2.jar (130 kB at 25 kB/s)
    Uploading: https://clojars.org/repo/net/mikera/core.matrix/0.60.2/core.matrix-0.60.2.pom
    Uploaded: https://clojars.org/repo/net/mikera/core.matrix/0.60.2/core.matrix-0.60.2.pom (5.9 kB at 2.0 kB/s)
    Downloading: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml
    Downloaded: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml (2.9 kB at 5.8 kB/s)
    Downloading: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml
    Downloaded: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml (2.9 kB at 5.9 kB/s)
    Uploading: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml
    Uploaded: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml (3.0 kB at 888 B/s)
    Uploading: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml
    [INFO] ------------------------------------------------------------------------
    [INFO] BUILD FAILURE
    [INFO] ------------------------------------------------------------------------
    [INFO] Total time: 51.337 s
    [INFO] Finished at: 2017-05-17T08:41:10+08:00
    [INFO] Final Memory: 31M/371M
    [INFO] ------------------------------------------------------------------------
    [ERROR] Failed to execute goal org.apache.maven.plugins:maven-deploy-plugin:2.7:deploy (default-deploy) on project core.matrix: Failed to deploy metadata: Could not transfer metadata net.mikera:core.matrix/maven-metadata.xml from/to clojars.org (https://clojars.org/repo): Access denied to: https://clojars.org/repo/net/mikera/core.matrix/maven-metadata.xml, ReasonPhrase: Forbidden - redeploying non-snapshots is not allowed (see https://git.io/v1IAs). -> [Help 1]
    [ERROR]
    [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
    [ERROR] Re-run Maven using the -X switch to enable full debug logging.
    [ERROR]
    [ERROR] For more information about the errors and possible solutions, please read the following articles:
    [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 03:34 min
[INFO] Finished at: 2017-05-17T08:41:11+08:00
[INFO] Final Memory: 12M/184M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-release-plugin:2.5.2:perform (default-cli) on project core.matrix: Maven execution failed, exit code: '1' -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

**Comment by mikera** [comment]
Additional thought: it would be helpful if Clojars didn't create / freeze a repository version when the deploy fails like this so that I can attempt redeploys with the same version

**Comment by tobias** [comment]
It works for me against a local test instance of clojars, but does indeed fail against the production server, which is frustrating. 

With regards to not freezing the version - as far as clojars knows, this deploy succeeded, since it looks like a standard deploy without classified artifacts. If the deploy had failed before uploading `maven-metadata.xml` (or while uploading the first copy), the deploy wouldn't be moved to the repo proper, and would be redeployable. But maven/aether gives us no indication that there are additional files, so we can't easily detect when there are additional artifacts coming.

**Comment by mikera** [comment]
Ah I see: that's annoying. Not an expert in Maven but it seems like a design flaw - shouldn't there be some kind of "complete deploy" message?

Perhaps there is a way to infer from the pom or the metadata that a new artifact should be expected (as opposed to redeploying an existing one?

**Comment by tobias** [comment]
I've got it working now (see https://clojars.org/repo/org/clojars/test4/core.matrix/0.60.11/). Turns out the old clojars process didn't exit when I deployed the new version, so it was running old code.

There's nothing in the pom or metadata that indicates that there are additional artifacts (other than trying to parse the maven-jar-plugin configuration in the pom, which would be unreliable). The proper answer is probably our own protocol for deployment, with changes to lein/boot to use that, along with a maven plugin, but I'd rather not go down that path, since it's more to maintain and more of a burden on users.
